go/cmd/dolt/commands/add.go

```diff
@@ -123,59 +138,558 @@ func (cmd AddCmd) Exec(ctx context.Context, commandStr string, args []string, dE
+		}
+		rows, err = sql.RowIterToRows(sqlCtx, rowIter)
+		if err != nil {
+			return nil, err
+		}
+		if len(rows) != 1 {
+			return nil, errors.New("Expected one row")
+		}
+		firstId, err := coerceToInt(rows[0][0])
+		if err != nil {
+			return nil, err
+		}
+		changeCounts[tableName].firstId = firstId
+		lastId, err := coerceToInt(rows[0][1])
+		if err != nil {
+			return nil, err
+		}
+		changeCounts[tableName].lastId = lastId
+
+	}
+
+	return changeCounts, nil
 }
 
-func toStageVErr(err error) errhand.VerboseError {
-	switch {
-	case doltdb.IsRootValUnreachable(err):
-		rt := doltdb.GetUnreachableRootType(err)
-		bdr := errhand.BuildDError("Unable to read %s.", rt.String())
-		bdr.AddCause(doltdb.GetUnreachableRootCause(err))
-		return bdr.Build()
+func coerceToInt(val interface{}) (int, error) {
+	switch v := val.(type) {
+	case int:
+		return v, nil
+	case int64:
+		return int(v), nil
+	case string:
+		i, err := strconv.Atoi(v)
+		if err != nil {
+			return 0, err
+		}
+		return i, nil
+	default:
+		return 0, errors.New("Expected int, int64 or string")
+	}
+}
+
+// queryForSingleChange queries the dolt_workspace_* table for the row with the given ID.
+func queryForSingleChange(sqlCtx *sql.Context, queryist cli.Queryist, tableName string, rowId int) (sql.Row, error) {
+	qry := fmt.Sprintf("SELECT * FROM dolt_workspace_%s WHERE ID = %d LIMIT 1", tableName, rowId)
+	_, rowIter, _, err := queryist.Query(sqlCtx, qry)
+	if err != nil {
+		return nil, err
+	}
+	rows, err := sql.RowIterToRows(sqlCtx, rowIter)
+	if err != nil {
+		return nil, err
+	}
+	if len(rows) == 0 || len(rows) > 1 {
+		return nil, errors.New("Expected exactly one row")
+	}
+	return rows[0], nil
+}
 
-	case actions.IsTblNotExist(err):
-		tbls := actions.GetTablesForError(err)
-		bdr := errhand.BuildDError("Some of the specified tables or docs were not found")
-		bdr.AddDetails("Unknown tables or docs: %v", tbls)
+func opHelp(_ *ishell.Context) {
+	help := `y - stage the current change
+n - do not stage the current change
+q - quit
+a - add all changes in this table
+d - do not stage any further changes in this table
+s - show summary of unstaged changes and start over
+? - show this help`
+	help = color.CyanString(help)
+	cli.Println(help)
+}
 
-		return bdr.Build()
+// patchState is the state for the patch workflow.
+type patchState struct {
+	sqlCtx                *sql.Context
+	queryist              cli.Queryist
+	tables                []string
+	changeCounts          map[string]*tablePatchInfo
+	currentTable          string
+	currentTableSchema    sql.Schema
+	currentTableLastRowId int
+	currentRowId          int
+	currentRow            sql.Row
+	err                   error
+}
 
-	case actions.IsTblInConflict(err) || actions.IsTblViolatesConstraints(err):
-		tbls := actions.GetTablesForError(err)
-		bdr := errhand.BuildDError("error: not all tables merged")
+// opYes stages the current change. "a" command.
+func (ps *patchState) opYes(c *ishell.Context) {
```

Comment:
(minor/style) The "op" prefix on these functions feels out of place to me. I know these are the functions you reference for the patch add operations in the shell, but it seems like more descriptive names like `stageCurrentChange` would convey more meaning than `opYes`.